cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -173,6 +222,172 @@ CUDF_HOST_DEVICE inline constexpr T shift(T const& val, scale_type const& scale)
+    // As none of these are representable by integers, we'll return the same value for all cases.
+    return cuda::std::tuple{false, INT_MIN, IntegralType(0)};
+  }
+
+  // Extract the exponent value: shift the bits down and subtract the bias.
+  using SignedIntegralType                       = cuda::std::make_signed_t<IntegralType>;
+  SignedIntegralType const shifted_exponent_bits = exponent_bits >> Constants::num_mantissa_bits;
+  int const exp2 =
+    shifted_exponent_bits - static_cast<SignedIntegralType>(Constants::exponent_bias);
+
+  // Extract the sign bit:
+  bool const is_positive = ((Constants::sign_mask & integer_rep) == 0);
+
+  // Extract the significand, setting the high bit for the understood 1/2
+  IntegralType const significand =
+    (integer_rep & Constants::mantissa_mask) | Constants::understood_bit_mask;
+
+  // Return the tuple of values
+  return cuda::std::tuple{is_positive, exp2, significand};
+}
+
+/** @brief Sets the sign and adds to the exponent of a floating point number.
+ *
+ * @tparam FloatingType Type of input floating point value.
+ * @param floating The floating value to set the components of.
+ * @param is_negative The sign bit to set for the floating point number.
+ * @param exp2 The power-of-2 to add to the floating point number.
+ * @return A floating point value with the mantissa of the original, but the added sign and
+ * exponent.
+ */
+template <typename FloatingType,
+          typename cuda::std::enable_if_t<cuda::std::is_floating_point_v<FloatingType>>* = nullptr>
+CUDF_HOST_DEVICE inline FloatingType add_sign_and_exp2(FloatingType floating,
+                                                       bool is_negative,
+                                                       int exp2)
+{
+  // Because IEEE-754 mandates rounding when casting an integer to floating point,
+  // the fastest way to set the mantissa is to cast, even though it uses the FPU.
+  // However, the fastest way to set the sign and exponent is still to do so manually,
+  // especially for doubles as bottlenecking on the FP64 pipeline is very painful.
+
+  // See comments in FloatingComponentConstants about the floating point format.
+  using Constants    = FloatingComponentConstants<FloatingType>;
+  using IntegralType = typename Constants::IntegralType;
+
+  // See comments in extract_components() about type-punning with a union.
+  // TODO: Use std::bit_cast in C++20
+  union UseBitCastInCpp20 {
+    IntegralType integer;
+    FloatingType floating;
+  };
+
+  // Convert floating point to integer
+  // We'll edit the integer in-place as we have to convert back to floating point at the end.
+  // Note that we do NOT use a reference or pointer to the union integer, as that IS undefined
+  // behavior in gcc.
+  UseBitCastInCpp20 conversion_union;
+  conversion_union.floating = floating;
+
+  // Set the sign bit
+  conversion_union.integer |= (IntegralType(is_negative) << Constants::sign_bit_index);
```

Comment:
this is UB, right? I think Jake recommended using memcpy to avoid breaking the strict aliasing rule.

Reply:
Not in gcc or clang.  But in CUDA I don't know, I'll look into that

Reply:
> It is undefined behavior to read from the member of the union that wasn't most recently written. Many compilers implement, as a non-standard language extension, the ability to read inactive members of a union.

https://en.cppreference.com/w/cpp/language/union

Reply:
Yes in the language, but the compilers we use define it.  However I didn't realize memcpy for this wasn't UB.  I've switched to using it instead.